File: configurations/i20-config/scripts/adc_monitor.py
from gda.device.detector import DetectorBase
from gda.epics import CAClient
import time

class adcmonitor(DetectorBase):

	def __init__(self,name,pvPrefix,readoutPV,columnName):
		self.setName(name)
		self.pvPrefix = pvPrefix
		self.ca = CAClient()
		self.readoutPV = readoutPV
		self.inputNames = []
		self.extraNames = [columnName]
		self.outputFormat = ["%.5g"]
		self.integrating = False
		self.collectionTime = 1.0

	def collectData(self):
		integrationTime = self.collectionTime
		self.ca.caput(self.pvPrefix + ":PERIOD",integrationTime)
		#self.ca.caput(self.pvPrefix + ":MODE","Trigger")
		self.integrating = True
		#self.ca.caputWait(self.pvPrefix + ":SOFTTRIGGER.VAL",1)
		self.ca.caput(self.pvPrefix + ":SOFTTRIGGER.VAL",1)
		time.sleep(0.25)

	def getStatus(self):
		if self.integrating == False:
			return 0
		
		triggering = str(self.ca.caget(self.pvPrefix + ":SOFTTRIGGER.VAL"))
		if triggering == "Done" or triggering == "0":
			self.integrating = False

		if self.integrating :
			return 1
		
		return 0

	def readout(self):
		return [float(self.ca.caget(self.readoutPV))]

	def getDescription(self):
		return " "

	def getDetectorID(self):
		return " "

	def getDetectorType(self):
		return " "
	
	def createsOwnFiles(self):
		return False
	
d1 = adcmonitor("d1","BL20I-DI-ADC-01","BL20I-DI-PHDGN-01:DIODE:I","d1")
d3plus = adcmonitor("d3plus","BL20I-DI-ADC-01","BL20I-DI-PHDGN-03:Y:PLUS:I","d3plus")
d3minus = adcmonitor("d3minus","BL20I-DI-ADC-01","BL20I-DI-PHDGN-03:Y:MINUS:I","d3minus")
d4  = adcmonitor("d4", "BL20I-DI-ADC-01","BL20I-DI-PHDGN-04:DIODE1:I","d4")
d5  = adcmonitor("d5", "BL20I-DI-ADC-01","BL20I-DI-PHDGN-05:DIODE:I","d5")
d9  = adcmonitor("d9", "BL20I-DI-ADC-02","BL20I-DI-PHDGN-09:DIODE:I","d9")
d10 = adcmonitor("d10","BL20I-DI-ADC-02","BL20I-DI-PHDGN-10:DIODE:I","d10")

File: plugins/uk.ac.gda.beamline.i20.server/src/uk/ac/gda/beamline/i20/scannable/ADCMonitor.java
/*-
 * Copyright © 2013 Diamond Light Source Ltd.
 *
 * This file is part of GDA.
 *
 * GDA is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License version 3 as published by the Free
 * Software Foundation.
 *
 * GDA is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along
 * with GDA. If not, see <http://www.gnu.org/licenses/>.
 */

package uk.ac.gda.beamline.i20.scannable;

import java.io.IOException;

import gda.device.Detector;
import gda.device.DeviceException;
import gda.device.detector.DetectorBase;
import gda.epics.LazyPVFactory;
import gda.epics.PV;
import gda.factory.FactoryException;

public class ADCMonitor extends DetectorBase {

	private static final long serialVersionUID = 1L;

	private final String pvPrefix;
	private final String readoutPVName;
	private final String columnName;
	private PV<Double> samplePeriodPV;
	private PV<ADCMonitorModes> modePV;
	private PV<Integer> triggerPV;
	private PV<Double> readoutPV;

	public ADCMonitor(String name, String pvPrefix, String readoutPV, String columnName) {
		setName(name);
		this.pvPrefix = pvPrefix;
		this.readoutPVName = readoutPV;
		this.columnName = columnName;
	}

	@Override
	public void configure() throws FactoryException {
		super.configure();
		setInputNames(new String[] {});
		setExtraNames(new String[] { columnName });
		setOutputFormat(new String[] { "%.5g" });
		samplePeriodPV = LazyPVFactory.newDoublePV(pvPrefix + ":PERIOD");
		modePV = LazyPVFactory.newEnumPV(pvPrefix + ":MODE", ADCMonitorModes.class);
		triggerPV = LazyPVFactory.newIntegerPV(pvPrefix + ":SOFTTRIGGER");
		readoutPV = LazyPVFactory.newDoublePV(readoutPVName);
	}

	@Override
	public void collectData() throws DeviceException {
		try {
			samplePeriodPV.putWait(getCollectionTime());
			triggerPV.putNoWait(1);
		} catch (IOException e) {
			throw new DeviceException(e.getMessage(), e);
		}
	}

	@Override
	public int getStatus() throws DeviceException {
		try {
			ADCMonitorModes currentMode = modePV.get();
			if (currentMode == ADCMonitorModes.Continuous || currentMode == ADCMonitorModes.Gate)
				return Detector.IDLE;
			if (triggerPV.get() == 0)
				return Detector.IDLE;
			return Detector.BUSY;
		} catch (IOException e) {
			throw new DeviceException(e.getMessage(), e);
		}
	}

	@Override
	public Object readout() throws DeviceException {
		try {
			return readoutPV.get();
		} catch (IOException e) {
			throw new DeviceException(e.getMessage(), e);
		}
	}

	@Override
	public void atScanStart() throws DeviceException {
		super.atScanStart();
		try {
			modePV.putWait(ADCMonitorModes.Trigger);
		} catch (IOException e) {
			throw new DeviceException(e.getMessage(), e);
		}
	}

	@Override
	public void atScanEnd() throws DeviceException {
		super.atScanEnd();
		try {
			modePV.putWait(ADCMonitorModes.Continuous);
		} catch (IOException e) {
			throw new DeviceException(e.getMessage(), e);
		}
	}

	@Override
	public void atCommandFailure() throws DeviceException {
		atScanEnd();
	}

	@Override
	public boolean createsOwnFiles() throws DeviceException {
		return false;
	}

}